cmd/gateclient/oauth.go

```diff
@@ -0,0 +1,85 @@
+package gateclient
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"net"
+	"net/http"
+	"os/exec"
+	"runtime"
+
+	"golang.org/x/oauth2"
+)
+
+const localWebServer = "localhost:8085"
+
+func startWebServer() (codeCh chan string, err error) {
+	listener, err := net.Listen("tcp", localWebServer)
+	if err != nil {
+		return nil, err
+	}
+	codeCh = make(chan string)
+
+	go http.Serve(listener, http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		// TODO: add handle of `error`
+		//error := r.FormValue("error")
+		code := r.FormValue("code")
+		codeCh <- code // send code to OAuth flow
+		listener.Close()
+		w.Header().Set("Content-Type", "text/plain")
+		fmt.Fprintf(w, "Received code: %v\r\nYou can now safely close this browser window.", code)
+	}))
+
+	return codeCh, nil
+}
+
+// getTokenFromWeb uses Config to request a Token.
+// It returns the retrieved Token.
+func getTokenFromWeb(output func(string), config *oauth2.Config, authURL string, opts ...oauth2.AuthCodeOption) (*oauth2.Token, error) {
+	codeCh, err := startWebServer()
+	if err != nil {
+		output("Unable to start a web server.")
+		return nil, err
+	}
+
+	err = openURL(authURL)
+	if err != nil {
+		log.Fatalf("Unable to open authorization URL in web server: %v", err)
+	} else {
+		output("Your browser has been opened to an authorization URL.\n" +
+			" This program will resume once authorization has been provided.\n")
+		output(authURL)
+	}
+
+	// Wait for the web server to get the code.
+	code := <-codeCh
+	return exchangeToken(config, code, opts...)
+}
+
+// Exchange the authorization code for an access token
+func exchangeToken(config *oauth2.Config, code string, opts ...oauth2.AuthCodeOption) (*oauth2.Token, error) {
+	tok, err := config.Exchange(context.Background(), code, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("Unable to retrieve token %v", err)
+	}
+	return tok, nil
+}
+
+// openURL opens a browser window to the specified location.
+// This code originally appeared at:
+//   http://stackoverflow.com/questions/10377243/how-can-i-launch-a-process-that-is-not-a-file-in-go
+func openURL(url string) error {
```

Comment:
Using a local web server or prompt is optional in the linked code and this seems quite robust. If local web server doesn't work then use the prompt. https://github.com/youtube/api-samples/blob/master/go/oauth2.go

I'm not sure what the failure modes are for the various `openURL()` commands, like `exec` hanging with no feedback. I've managed to misconfigure Linux for `xdg-open` more than once :sweat_smile: 

If an error is returned we bail out at L48 with `log.Fatalf` and there is no way for a user to progress unless they fix the exec issue.

I wonder if we might want to do something like:
a) fall back to prompt for code?
b) make `localWebServer` selectable via spin config (default enabled?) similer to https://github.com/youtube/api-samples/blob/07263305b59a7c3275bc7e925f9ce6cabf774022/go/oauth2.go#L44 

WDYT?